### ISSUE TITLE
string() doesn't work in pseudo-elements

### ISSUE THREAD

**Issue by Smylers** [Kozea/WeasyPrint#723]
The `string(foo)` function returns the contents of the specified named string in page margin boxes, but just returns the empty string in the main page contents — specifically in `:before` and `:after` pseudo-elements, which are the other place where the `content:` property can be set to arbitrary text.

Demo:

````html
<!DOCTYPE html>
<title>content: string Check</title>
<style>
body { string-set: title 'flrbbbb' }
@page { @top-left  { content: '[' string(title) ']' } }
p:after            { content: '{' string(title) '}' }
</style>

<p>After: 
````

The page header shows the expected “[flrbbbb]”, but the paragraph just says “After: {}”, with nothing in the curly brackets.

The [draft spec](https://drafts.csswg.org/css-content-3/#named-strings) says “Named strings are a convenient way to pull metadata out of the document for insertion into headers and footers” — but I can't find anything that says they should _only_ work there, rather than everywhere that `content:` can be used.

(However, I haven't yet found a reason for doing this in a real document; I encountered the limitation simply while trying to construct a minimal demo for Issue #722.)

**Comment by Tontyna** [comment]
@SimonSapin's statement on [GCPM and string()](#174#issuecomment-76711593) still holds.

In the spec the function is only mentioned when they talk about content in the page (margins), so my understanding  of the spec is that `string()` is only supported in `@page` context. That's why you can blame me for this [comment line](https://github.com/Kozea/WeasyPrint/blob/d2169a75f51b8b3ff1520318cf6081b42dc03307/weasyprint/formatting_structure/build.py#L325) and excluding condition in `build.py`.

But of course, @Smylers is right, too: Nowhere in the spec the use of `string()` outside of `@page` is forbidden explicitly -- though I have no idea how to implement the (page-related!) specifiers `first | start | last | first-except` in the document correctly.

@liZe you're sure that this is a bug?

**Comment by Smylers** [comment]
> @SimonSapin's statement on [GCPM and string()](https://github.com/Kozea/WeasyPrint/issues/174#issuecomment-76711593) still holds.

Ah, I hadn't seen that. Thanks.

> In the spec the function is only mentioned when they talk about content in the page (margins), so my understanding of the spec is that `string()` is only supported in `@page` context.

OK. In that case, please could it actually be invalid, like any other unknown function is?

Changing `string()` to, say, `stringX()` in the above demo causes the pseudo-element's content to be completely empty (rather than ‘{}’ from the constants either side of the ignored function), and the emission of this warning:

````
WARNING: Ignored `content: "{" stringX(title) "}" ` at 4:22, invalid value.
````

If `string()` in pseudo-elements also did that, it'd be clearer that it isn't intended to work.

**Comment by Tontyna** [comment]
Emitting a WARNING is no problem, only the `at col:line` part telling where to look in the css file is -- @liZe correct me if I'm wrong -- impossible.

**Comment by Tontyna** [comment]
Ah, and I remember there was a discussion whether to throw away the whole `content` or only the invalid `string()` part

**Comment by liZe** [comment]
> @SimonSapin's statement on [GCPM and string()](#174#issuecomment-76711593) still holds.

True, I didn't remember that. But the feature may be useful, and nothing seems to prevent this in the spec.

> For example, it’s not clear to me whether the string() function of the content property is valid in contexts other than on page-margin boxes (e.g. ::before or ::after pseudo-elements) and what the second argument means in this case.

There's no logical problem now with the second argument, it should behave just as it does when `string` is used in page margins.

> OK. In that case, please could it actually be invalid, like any other unknown function is?

Yes, we should.

> only the `at col:line` part telling where to look in the css file is -- @liZe correct me if I'm wrong -- impossible.

It's actually possible, it's even already done for most of CSS errors :smile:. What's impossible (without lxml) is to get the position in HTML files.

> Ah, and I remember there was a discussion whether to throw away the whole `content` or only the invalid `string()` part

We should invalidate the whole rule if we decide that `string` is not allowed (I hope that it's what I said in the previous discussion).

I'll invalidate the rule and keep this issue open to add the feature.

**Comment by Tontyna** [comment]
> It's actually possible,
> I'll invalidate the rule

@liZe Looking forward how you'll implement that :smirk: 

Reason-for-smirk:
 `content()` in `css.validation.properies` resp. `get_content_list()` in `css.utils` has no idea to what kind of css rule (page, pseudo element, real element) the `content` property belongs.

We know the element (but not the `col:line` where it's css is defined) in `computed_values.py` --  I recall a version of `computed_values.py` where I discarded parts of the `content_list` depending on the element , but this "validation" was removed by ... authority ... and moved to `compute_content_list()` in `build.py`

**Comment by liZe** [comment]
> @liZe Looking forward how you'll implement that

:laughing:

**Comment by liZe** [comment]
> We should invalidate the whole rule if we decide that `string` is not allowed (I hope that it's what I said in the previous discussion).

Of course, that's what we should do if we had the possibility to know if the function is called in the page margin box or elsewhere.

> `content()` in `css.validation.properies` resp. `get_content_list()` in `css.utils` has no idea to what kind of css rule (page, pseudo element, real element) the `content` property belongs.

Well… You're right.

A solution would be to have a list of validation functions for each context (page, page margin, page content). It also would be useful to reject properties that are not allowed in a particular context (here are the properties allowed [for page](https://drafts.csswg.org/css-page-3/#page-property-list) and [for page margin](https://drafts.csswg.org/css-page-3/#margin-property-list)).

> but this "validation" was removed by ... authority ... and moved to `compute_content_list()` in `build.py`

If we decide that `string()` is not allowed in the rule, then we have to refuse the whole rule during validation, not during value computation, to allow the use of a fallback rule defined earlier.

For example:
``` css
p::before {
  content: 'lol';
  content: '[' string(title) ']';
}
```
should keep 'lol' if we decide to refuse `string()`, and that's what's done during validation. After validation it's too late to discard the whole second rule.

The current implementation (and 36ef44c keeps it with just a warning message) allows `string()` but makes it behave like an empty string.

**Comment by Tontyna** [comment]
Ah, stupid me: As @liZe pointed out the rule's context -- page, page margin, page content -- should be sufficient for enhanced css validation and this context is well-known in the rule.
